File: cli-proxy-logger/src/kernel/links.js
// Share-link parsing -> normalized outbound spec (pure, zero-dependency).
//
// Turns common proxy share links into a kernel-agnostic descriptor that the
// xray / sing-box config renderers consume. Supported schemes:
//   vmess://      (v2rayN base64-JSON form)
//   vless://      (uuid@host:port?query#name)
//   trojan://     (password@host:port?query#name)
//   ss://         (SIP002: base64(method:pass)@host:port, or fully base64'd)
//   hysteria2://  (also hy2://) password@host:port?query#name
//   tuic://       (uuid:password@host:port?query#name)
//
// The returned spec shape (fields present depend on protocol):
//   {
//     type, tag, server, port,
//     uuid, alterId, security,           // vmess
//     flow,                              // vless
//     password,                          // trojan / ss / hysteria2 / tuic
//     method,                            // shadowsocks
//     obfs: { type, password },          // hysteria2
//     congestionControl, alpn: [],       // tuic
//     tls: { enabled, serverName, insecure, alpn, fingerprint,
//            reality: { publicKey, shortId } | null } | null,
//     transport: { type, path, host, serviceName, headers } | null,
//     name,
//   }

const ADVANCED_SCHEMES = new Set([
  'vmess', 'vless', 'trojan', 'ss', 'shadowsocks', 'hysteria2', 'hy2', 'tuic',
]);

// Is this a link that needs a kernel (vs. a plain http/socks proxy)?
export function isKernelScheme(raw) {
  if (!raw || typeof raw !== 'string') return false;
  const m = /^([a-z0-9]+):\/\//i.exec(raw.trim());
  return !!m && ADVANCED_SCHEMES.has(m[1].toLowerCase());
}

function b64decode(s) {
  let str = String(s).trim().replace(/-/g, '+').replace(/_/g, '/');
  while (str.length % 4) str += '=';
  return Buffer.from(str, 'base64').toString('utf8');
}

// Build the tls block from URL query params shared by vless/trojan/tuic/hy2.
function tlsFromQuery(q, fallbackSni) {
  const security = (q.get('security') || '').toLowerCase();
  const sni = q.get('sni') || q.get('peer') || fallbackSni || '';
  const alpn = q.get('alpn') ? q.get('alpn').split(',').map((s) => s.trim()).filter(Boolean) : [];
  const fp = q.get('fp') || '';
  const insecure = q.get('allowInsecure') === '1' || q.get('insecure') === '1' || q.get('allow_insecure') === '1';
  const isReality = security === 'reality';
  const enabled = security === 'tls' || security === 'xtls' || isReality;
  if (!enabled && !sni && !alpn.length && !fp && !insecure) return null;
  const tls = { enabled: enabled || isReality, serverName: sni, insecure, alpn, fingerprint: fp, reality: null };
  if (isReality) {
    tls.reality = {
      publicKey: q.get('pbk') || '',
      shortId: q.get('sid') || '',
      spiderX: q.get('spx') || '',
    };
  }
  return tls;
}

// Build the transport block from URL query params (vless/trojan v2ray style).
function transportFromQuery(q) {
  const type = (q.get('type') || 'tcp').toLowerCase();
  if (type === 'tcp' || type === '' || type === 'none' || type === 'raw') return null;
  const host = q.get('host') || '';
  const path = q.get('path') || '';
  const serviceName = q.get('serviceName') || q.get('servicename') || path;
  return {
    type: type === 'h2' ? 'http' : type,
    path,
    host,
    serviceName,
    headers: host ? { Host: host } : {},
  };
}

function parseVmess(raw) {
  // vmess://<base64 json>
  const body = raw.slice('vmess://'.length);
  let obj;
  try {
    obj = JSON.parse(b64decode(body));
  } catch {
    throw new Error('vmess: not a base64-encoded JSON link');
  }
  const net = (obj.net || 'tcp').toLowerCase();
  const tlsOn = String(obj.tls || '').toLowerCase() === 'tls';
  const host = obj.host || '';
  const transport = (net !== 'tcp' && net !== '')
    ? {
      type: net === 'h2' ? 'http' : net,
      path: obj.path || '',
      host,
      serviceName: obj.path || '',
      headers: host ? { Host: host } : {},
    }
    : null;
  return {
    type: 'vmess',
    tag: 'proxy',
    server: obj.add,
    port: Number.parseInt(obj.port, 10),
    uuid: obj.id,
    alterId: Number.parseInt(obj.aid || 0, 10) || 0,
    security: obj.scy || obj.security || 'auto',
    tls: tlsOn
      ? { enabled: true, serverName: obj.sni || host || '', insecure: false, alpn: obj.alpn ? String(obj.alpn).split(',').filter(Boolean) : [], fingerprint: obj.fp || '', reality: null }
      : null,
    transport,
    name: obj.ps || '',
  };
}

function parseUserHostLink(raw, type) {
  // Generic uuid|password @ host:port ?query #name
  const u = new URL(raw);
  const q = u.searchParams;
  const port = Number.parseInt(u.port, 10);
  const spec = {
    type,
    tag: 'proxy',
    server: decodeURIComponent(u.hostname),
    port,
    name: u.hash ? decodeURIComponent(u.hash.slice(1)) : '',
    tls: tlsFromQuery(q, u.hostname),
    transport: transportFromQuery(q),
  };
  if (type === 'vless') {
    spec.uuid = decodeURIComponent(u.username);
    spec.flow = q.get('flow') || '';
    spec.encryption = q.get('encryption') || 'none';
  } else if (type === 'trojan') {
    spec.password = decodeURIComponent(u.username);
  }
  return spec;
}

function parseShadowsocks(raw) {
  // SIP002: ss://base64(method:password)@host:port#name
  // Legacy:  ss://base64(method:password@host:port)#name
  const hashIdx = raw.indexOf('#');
  const name = hashIdx >= 0 ? decodeURIComponent(raw.slice(hashIdx + 1)) : '';
  let body = (hashIdx >= 0 ? raw.slice(0, hashIdx) : raw).slice('ss://'.length);
  const qIdx = body.indexOf('?');
  let query = '';
  if (qIdx >= 0) { query = body.slice(qIdx + 1); body = body.slice(0, qIdx); }

  let method; let password; let server; let port;
  if (body.includes('@')) {
    const at = body.lastIndexOf('@');
    const userinfo = body.slice(0, at);
    const hostport = body.slice(at + 1);
    const decoded = /[:]/.test(userinfo) && !/^[A-Za-z0-9+/_=-]+$/.test(userinfo)
      ? userinfo
      : b64decode(userinfo);
    const ci = decoded.indexOf(':');
    method = decoded.slice(0, ci);
    password = decoded.slice(ci + 1);
    const hp = hostport.lastIndexOf(':');
    server = hostport.slice(0, hp);
    port = Number.parseInt(hostport.slice(hp + 1), 10);
  } else {
    const decoded = b64decode(body);
    const at = decoded.lastIndexOf('@');
    const cred = decoded.slice(0, at);
    const hostport = decoded.slice(at + 1);
    const ci = cred.indexOf(':');
    method = cred.slice(0, ci);
    password = cred.slice(ci + 1);
    const hp = hostport.lastIndexOf(':');
    server = hostport.slice(0, hp);
    port = Number.parseInt(hostport.slice(hp + 1), 10);
  }
  const q = new URLSearchParams(query);
  return {
    type: 'shadowsocks',
    tag: 'proxy',
    server,
    port,
    method,
    password,
    plugin: q.get('plugin') || '',
    tls: null,
    transport: null,
    name,
  };
}

function parseHysteria2(raw) {
  const u = new URL(raw.replace(/^hy2:\/\//, 'hysteria2://'));
  const q = u.searchParams;
  const tls = tlsFromQuery(q, u.hostname) || { enabled: true, serverName: u.hostname, insecure: false, alpn: [], fingerprint: '', reality: null };
  tls.enabled = true;
  return {
    type: 'hysteria2',
    tag: 'proxy',
    server: decodeURIComponent(u.hostname),
    port: Number.parseInt(u.port, 10) || 443,
    password: decodeURIComponent(u.username || u.password || ''),
    obfs: q.get('obfs') ? { type: q.get('obfs'), password: q.get('obfs-password') || q.get('obfs_password') || '' } : null,
    tls,
    transport: null,
    name: u.hash ? decodeURIComponent(u.hash.slice(1)) : '',
  };
}

function parseTuic(raw) {
  const u = new URL(raw);
  const q = u.searchParams;
  const tls = tlsFromQuery(q, u.hostname) || { enabled: true, serverName: u.hostname, insecure: false, alpn: [], fingerprint: '', reality: null };
  tls.enabled = true;
  return {
    type: 'tuic',
    tag: 'proxy',
    server: decodeURIComponent(u.hostname),
    port: Number.parseInt(u.port, 10) || 443,
    uuid: decodeURIComponent(u.username || ''),
    password: decodeURIComponent(u.password || ''),
    congestionControl: q.get('congestion_control') || q.get('congestion') || '',
    udpRelayMode: q.get('udp_relay_mode') || '',
    alpn: q.get('alpn') ? q.get('alpn').split(',').map((s) => s.trim()).filter(Boolean) : [],
    tls,
    transport: null,
    name: u.hash ? decodeURIComponent(u.hash.slice(1)) : '',
  };
}

// Parse a share link into a normalized outbound spec. Throws on malformed input.
export function parseShareLink(raw) {
  if (!raw || typeof raw !== 'string') throw new Error('empty link');
  const link = raw.trim();
  const scheme = (/^([a-z0-9]+):\/\//i.exec(link) || [])[1];
  if (!scheme) throw new Error('link missing scheme');
  switch (scheme.toLowerCase()) {
    case 'vmess': return parseVmess(link);
    case 'vless': return parseUserHostLink(link, 'vless');
    case 'trojan': return parseUserHostLink(link, 'trojan');
    case 'ss':
    case 'shadowsocks': return parseShadowsocks(link);
    case 'hysteria2':
    case 'hy2': return parseHysteria2(link);
    case 'tuic': return parseTuic(link);
    default: throw new Error(`unsupported link scheme: ${scheme}`);
  }
}

// Protocols only sing-box can handle (xray-core has no native hy2/tuic outbound).
export const SINGBOX_ONLY = new Set(['hysteria2', 'tuic']);

File: cli-proxy-logger-py/cli_proxy_logger/kernel/links.py
"""Share-link parsing -> normalized outbound spec (pure, zero-dependency).

Port of the Node ``src/kernel/links.js``. Turns common proxy share links into a
kernel-agnostic dict that the xray / sing-box config renderers consume.
Supported schemes::

    vmess://      (v2rayN base64-JSON form)
    vless://      (uuid@host:port?query#name)
    trojan://     (password@host:port?query#name)
    ss://         (SIP002: base64(method:pass)@host:port, or fully base64'd)
    hysteria2://  (also hy2://) password@host:port?query#name
    tuic://       (uuid:password@host:port?query#name)
"""

import base64
import json
import re
from urllib.parse import parse_qs, unquote, urlparse

ADVANCED_SCHEMES = {
    "vmess", "vless", "trojan", "ss", "shadowsocks", "hysteria2", "hy2", "tuic",
}

# Protocols only sing-box can handle (xray-core has no native hy2/tuic outbound).
SINGBOX_ONLY = {"hysteria2", "tuic"}

_SCHEME_RE = re.compile(r"^([a-z0-9]+)://", re.IGNORECASE)


def is_kernel_scheme(raw):
    """Is this a link that needs a kernel (vs. a plain http/socks proxy)?"""
    if not isinstance(raw, str):
        return False
    m = _SCHEME_RE.match(raw.strip())
    return bool(m) and m.group(1).lower() in ADVANCED_SCHEMES


def _b64decode(s):
    txt = str(s).strip().replace("-", "+").replace("_", "/")
    txt += "=" * (-len(txt) % 4)
    return base64.b64decode(txt).decode("utf-8")


class _Query:
    """Tiny wrapper over parse_qs giving JS URLSearchParams.get semantics."""

    def __init__(self, query):
        self._d = parse_qs(query, keep_blank_values=True)

    def get(self, key, default=""):
        vals = self._d.get(key)
        return vals[0] if vals else default


def _tls_from_query(q, fallback_sni):
    security = (q.get("security") or "").lower()
    sni = q.get("sni") or q.get("peer") or fallback_sni or ""
    alpn_raw = q.get("alpn")
    alpn = [s.strip() for s in alpn_raw.split(",") if s.strip()] if alpn_raw else []
    fp = q.get("fp") or ""
    insecure = (q.get("allowInsecure") == "1" or q.get("insecure") == "1"
                or q.get("allow_insecure") == "1")
    is_reality = security == "reality"
    enabled = security in ("tls", "xtls") or is_reality
    if not enabled and not sni and not alpn and not fp and not insecure:
        return None
    tls = {
        "enabled": enabled or is_reality,
        "serverName": sni,
        "insecure": insecure,
        "alpn": alpn,
        "fingerprint": fp,
        "reality": None,
    }
    if is_reality:
        tls["reality"] = {
            "publicKey": q.get("pbk") or "",
            "shortId": q.get("sid") or "",
            "spiderX": q.get("spx") or "",
        }
    return tls


def _transport_from_query(q):
    type_ = (q.get("type") or "tcp").lower()
    if type_ in ("tcp", "", "none", "raw"):
        return None
    host = q.get("host") or ""
    path = q.get("path") or ""
    service_name = q.get("serviceName") or q.get("servicename") or path
    return {
        "type": "http" if type_ == "h2" else type_,
        "path": path,
        "host": host,
        "serviceName": service_name,
        "headers": {"Host": host} if host else {},
    }


def _parse_vmess(raw):
    body = raw[len("vmess://"):]
    try:
        obj = json.loads(_b64decode(body))
    except (ValueError, json.JSONDecodeError) as exc:
        raise ValueError("vmess: not a base64-encoded JSON link") from exc
    net = str(obj.get("net") or "tcp").lower()
    tls_on = str(obj.get("tls") or "").lower() == "tls"
    host = obj.get("host") or ""
    transport = None
    if net not in ("tcp", ""):
        transport = {
            "type": "http" if net == "h2" else net,
            "path": obj.get("path") or "",
            "host": host,
            "serviceName": obj.get("path") or "",
            "headers": {"Host": host} if host else {},
        }
    tls = None
    if tls_on:
        alpn = obj.get("alpn")
        tls = {
            "enabled": True,
            "serverName": obj.get("sni") or host or "",
            "insecure": False,
            "alpn": [s for s in str(alpn).split(",") if s] if alpn else [],
            "fingerprint": obj.get("fp") or "",
            "reality": None,
        }
    return {
        "type": "vmess",
        "tag": "proxy",
        "server": obj.get("add"),
        "port": int(obj.get("port")),
        "uuid": obj.get("id"),
        "alterId": int(obj.get("aid") or 0),
        "security": obj.get("scy") or obj.get("security") or "auto",
        "tls": tls,
        "transport": transport,
        "name": obj.get("ps") or "",
    }


def _parse_user_host_link(raw, type_):
    u = urlparse(raw)
    q = _Query(u.query)
    spec = {
        "type": type_,
        "tag": "proxy",
        "server": unquote(u.hostname or ""),
        "port": u.port,
        "name": unquote(u.fragment) if u.fragment else "",
        "tls": _tls_from_query(q, u.hostname or ""),
        "transport": _transport_from_query(q),
    }
    if type_ == "vless":
        spec["uuid"] = unquote(u.username or "")
        spec["flow"] = q.get("flow") or ""
        spec["encryption"] = q.get("encryption") or "none"
    elif type_ == "trojan":
        spec["password"] = unquote(u.username or "")
    return spec


def _parse_shadowsocks(raw):
    hash_idx = raw.find("#")
    name = unquote(raw[hash_idx + 1:]) if hash_idx >= 0 else ""
    body = (raw[:hash_idx] if hash_idx >= 0 else raw)[len("ss://"):]
    q_idx = body.find("?")
    query = ""
    if q_idx >= 0:
        query = body[q_idx + 1:]
        body = body[:q_idx]

    if "@" in body:
        at = body.rfind("@")
        userinfo = body[:at]
        hostport = body[at + 1:]
        if ":" in userinfo and not re.fullmatch(r"[A-Za-z0-9+/_=-]+", userinfo):
            decoded = userinfo
        else:
            decoded = _b64decode(userinfo)
    else:
        decoded_all = _b64decode(body)
        at = decoded_all.rfind("@")
        decoded = decoded_all[:at]
        hostport = decoded_all[at + 1:]
    ci = decoded.find(":")
    method = decoded[:ci]
    password = decoded[ci + 1:]
    hp = hostport.rfind(":")
    server = hostport[:hp]
    port = int(hostport[hp + 1:])

    q = _Query(query)
    return {
        "type": "shadowsocks",
        "tag": "proxy",
        "server": server,
        "port": port,
        "method": method,
        "password": password,
        "plugin": q.get("plugin") or "",
        "tls": None,
        "transport": None,
        "name": name,
    }


def _parse_hysteria2(raw):
    u = urlparse(re.sub(r"^hy2://", "hysteria2://", raw))
    q = _Query(u.query)
    tls = _tls_from_query(q, u.hostname or "") or {
        "enabled": True, "serverName": u.hostname or "", "insecure": False,
        "alpn": [], "fingerprint": "", "reality": None,
    }
    tls["enabled"] = True
    obfs = None
    if q.get("obfs"):
        obfs = {"type": q.get("obfs"),
                "password": q.get("obfs-password") or q.get("obfs_password") or ""}
    return {
        "type": "hysteria2",
        "tag": "proxy",
        "server": unquote(u.hostname or ""),
        "port": u.port or 443,
        "password": unquote(u.username or u.password or ""),
        "obfs": obfs,
        "tls": tls,
        "transport": None,
        "name": unquote(u.fragment) if u.fragment else "",
    }


def _parse_tuic(raw):
    u = urlparse(raw)
    q = _Query(u.query)
    tls = _tls_from_query(q, u.hostname or "") or {
        "enabled": True, "serverName": u.hostname or "", "insecure": False,
        "alpn": [], "fingerprint": "", "reality": None,
    }
    tls["enabled"] = True
    alpn_raw = q.get("alpn")
    return {
        "type": "tuic",
        "tag": "proxy",
        "server": unquote(u.hostname or ""),
        "port": u.port or 443,
        "uuid": unquote(u.username or ""),
        "password": unquote(u.password or ""),
        "congestionControl": q.get("congestion_control") or q.get("congestion") or "",
        "udpRelayMode": q.get("udp_relay_mode") or "",
        "alpn": [s.strip() for s in alpn_raw.split(",") if s.strip()] if alpn_raw else [],
        "tls": tls,
        "transport": None,
        "name": unquote(u.fragment) if u.fragment else "",
    }


def parse_share_link(raw):
    """Parse a share link into a normalized outbound spec. Raises on bad input."""
    if not isinstance(raw, str) or not raw:
        raise ValueError("empty link")
    link = raw.strip()
    m = _SCHEME_RE.match(link)
    if not m:
        raise ValueError("link missing scheme")
    scheme = m.group(1).lower()
    if scheme == "vmess":
        return _parse_vmess(link)
    if scheme == "vless":
        return _parse_user_host_link(link, "vless")
    if scheme == "trojan":
        return _parse_user_host_link(link, "trojan")
    if scheme in ("ss", "shadowsocks"):
        return _parse_shadowsocks(link)
    if scheme in ("hysteria2", "hy2"):
        return _parse_hysteria2(link)
    if scheme == "tuic":
        return _parse_tuic(link)
    raise ValueError(f"unsupported link scheme: {scheme}")
